From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrPowerGamerBR <git@mrpowergamerbr.com>
Date: Sun, 25 May 2025 21:39:32 -0300
Subject: [PATCH] SparklyPaper Parallel World Ticking MSPT Tracking


diff --git a/src/main/java/io/papermc/paper/command/MSPTCommand.java b/src/main/java/io/papermc/paper/command/MSPTCommand.java
index 04b8097b147350f97fdee16c0b8aa78a87fddd33..391e864a359af538959536a78a91f2cf0425f0bb 100644
--- a/src/main/java/io/papermc/paper/command/MSPTCommand.java
+++ b/src/main/java/io/papermc/paper/command/MSPTCommand.java
@@ -82,6 +82,49 @@ public final class MSPTCommand extends Command {
                 )
             )
         );
+
+        // SparklyPaper start - track world's MSPT
+        if (net.serlith.fish.FishConfig.ASYNC.WORLD_TICKING._ENABLED) {
+            sender.sendMessage(text());
+            sender.sendMessage(text().content("World tick times ").color(GOLD)
+                .append(text().color(YELLOW)
+                    .append(
+                        text("("),
+                        text("avg", GRAY),
+                        text("/"),
+                        text("min", GRAY),
+                        text("/"),
+                        text("max", GRAY),
+                        text(")")
+                    )
+                ).append(
+                    text(" from last 5s"),
+                    text(",", GRAY),
+                    text(" 10s"),
+                    text(",", GRAY),
+                    text(" 1m"),
+                    text(":", YELLOW)
+                )
+            );
+            for (net.minecraft.server.level.ServerLevel serverLevel : server.getAllLevels()) {
+                List<Component> worldTimes = new ArrayList<>();
+                worldTimes.addAll(eval(serverLevel.fish$tickTimes5s));
+                worldTimes.addAll(eval(serverLevel.fish$tickTimes10s));
+                worldTimes.addAll(eval(serverLevel.fish$tickTimes60s));
+
+                sender.sendMessage(text().content("◴ " + serverLevel.getWorld().getName() + ": ").color(GOLD)
+                    .append(text().color(GRAY)
+                        .append(
+                            worldTimes.get(0), SLASH, worldTimes.get(1), SLASH, worldTimes.get(2), text(", ", YELLOW),
+                            worldTimes.get(3), SLASH, worldTimes.get(4), SLASH, worldTimes.get(5), text(", ", YELLOW),
+                            worldTimes.get(6), SLASH, worldTimes.get(7), SLASH, worldTimes.get(8)
+                        )
+                    )
+                );
+            }
+        }
+        // SparklyPaper end
+
         return true;
     }
 
