From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Biquaternions <biquaternions@serlith.net>
Date: Sat, 19 Apr 2025 21:18:45 -0500
Subject: [PATCH] Fish Added Config And Command


diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index 49f9e3aa446372389442eee835a9c829b23101b8..699c6db7ce21f301dd1eecb44be2a382a18257c2 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -110,6 +110,7 @@ public class Main {
                 JvmProfiler.INSTANCE.start(Environment.SERVER);
             }
 
+            me.biquaternions.fish.FishConfig.init(); // Fish - Config and command
             io.papermc.paper.plugin.PluginInitializerManager.load(options); // Paper
             Bootstrap.bootStrap();
             Bootstrap.validate();
diff --git a/net/minecraft/server/ReloadableServerResources.java b/net/minecraft/server/ReloadableServerResources.java
index 6db3af5861a25cff70fc940d1cd467ea81d4c29f..6d8a08ac361a7caa118a9ff4e6e3085adebda9ec 100644
--- a/net/minecraft/server/ReloadableServerResources.java
+++ b/net/minecraft/server/ReloadableServerResources.java
@@ -49,6 +49,7 @@ public class ReloadableServerResources {
         io.papermc.paper.command.PaperCommands.registerCommands(); // Paper
         gg.pufferfish.pufferfish.PufferfishCommand.init(); // Pufferfish - Config and command
         gg.pufferfish.pufferfish.flare.FlareCommand.init(); // Pufferfish - Config and command
+        me.biquaternions.fish.FishCommand.init(); // Fish - Config and command
         this.advancements = new ServerAdvancementManager(loadingContext);
         this.functionLibrary = new ServerFunctionLibrary(functionCompilationPermissions, this.commands.getDispatcher());
     }
