From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Biquaternions <biquaternions@serlith.net>
Date: Sat, 18 Jul 2026 15:23:38 -0500
Subject: [PATCH] Update


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index cff4246db3dd278de34597e3554611cd9efd28b8..4130999a89d1fad0e45485e10ad04184cf05adde 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -463,6 +463,10 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     @Override
     public final void moonrise$executeMidTickTasks() {
+        if (me.biquaternions.fish.FishConfig.ASYNC.WORLD_TICKING._ENABLED && me.biquaternions.fish.async.thread.WorldTickThread.isWorldTickThread()) {
+            me.biquaternions.fish.async.AsyncWorldTicking.executeMidTickTasks();
+            return;
+        }
         final long startTime = System.nanoTime();
         if ((startTime - this.lastMidTickExecute) <= CHUNK_TASK_QUEUE_BACKOFF_MIN_TIME || (startTime - this.lastMidTickExecuteFailure) <= TASK_EXECUTION_FAILURE_BACKOFF) {
             // it's shown to be bad to constantly hit the queue (chunk loads slow to a crawl), even if no tasks are executed.
