From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Biquaternions <biquaternions@serlith.net>
Date: Sat, 9 Aug 2025 17:35:49 -0500
Subject: [PATCH] Fish Rebrand


diff --git a/src/main/java/gg/pufferfish/pufferfish/PufferfishVersionFetcher.java b/src/main/java/gg/pufferfish/pufferfish/PufferfishVersionFetcher.java
index 506b6a32c2be1ff599c3205f00b34870001f9c46..ca0783676874ecc33b90dcc9f0eaf7d5c4837fb9 100644
--- a/src/main/java/gg/pufferfish/pufferfish/PufferfishVersionFetcher.java
+++ b/src/main/java/gg/pufferfish/pufferfish/PufferfishVersionFetcher.java
@@ -35,12 +35,12 @@ import static net.minecraft.server.MinecraftServer.COMPONENT_LOGGER;
 @NullMarked
 public class PufferfishVersionFetcher extends PaperVersionFetcher {
 
-    private static final Logger LOGGER = LoggerFactory.getLogger(PufferfishVersionFetcher.class.getSimpleName());
+    private static final Logger LOGGER = LoggerFactory.getLogger("FishVersionFetcher"); // Fish - Rebrand
     private static final long CACHE_TIME = TimeUnit.MINUTES.toMillis(30);
 
-    private static final String REPOSITORY = "SerlithNetwork/Puffernot";
-    private static final String DOWNLOAD_PAGE = "https://github.com/SerlithNetwork/Puffernot/releases";
-    private static final TextColor COLOR_SECONDARY = TextColor.color(0x87, 0xDD, 0xFF);
+    private static final String REPOSITORY = "Biquaternions/Fish"; // Fish - Rebrand
+    private static final String DOWNLOAD_PAGE = "https://github.com/Biquaternions/Fish/releases"; // Fish - Rebrand
+    private static final TextColor COLOR_SECONDARY = TextColor.color(0xA1, 0xF9, 0xFF); // Fish - Rebrand
 
     private static final int DISTANCE_ERROR = -1;
     private static final int DISTANCE_UNKNOWN = -2;
@@ -103,7 +103,7 @@ public class PufferfishVersionFetcher extends PaperVersionFetcher {
 
     private static int fetchDistanceFromSiteApi(final int actionsBuild) {
         try {
-            final URL buildsUrl = URI.create("https://version.serlith.net/api/v1/project/fetch/pufferfish/" + PufferfishVersionFetcher.BUILD_INFO.minecraftVersionId() + "/latest/build").toURL();
+            final URL buildsUrl = URI.create("https://version.serlith.net/api/v1/project/fetch/fish/" + PufferfishVersionFetcher.BUILD_INFO.minecraftVersionId() + "/latest/build").toURL(); // Fish - Rebrand
             final HttpURLConnection connection = (HttpURLConnection) buildsUrl.openConnection();
             connection.setConnectTimeout(5000);
             connection.setReadTimeout(5000);
@@ -118,7 +118,7 @@ public class PufferfishVersionFetcher extends PaperVersionFetcher {
                 final int latest = json.getAsInt();
                 return Math.max(latest - actionsBuild, 0);
             } catch (final JsonSyntaxException | MalformedJsonException ex) {
-                LOGGER.error("Error parsing json from Pufferfish's downloads API", ex);
+                LOGGER.error("Error parsing json from Fish's downloads API", ex); // Fish - Rebrand
                 return DISTANCE_ERROR;
             } catch (final SocketTimeoutException ex) {
                 return DISTANCE_ERROR;
diff --git a/src/main/java/gg/pufferfish/pufferfish/compat/ServerConfigurations.java b/src/main/java/gg/pufferfish/pufferfish/compat/ServerConfigurations.java
index b98ae2e7dea76df000f3071a27841042e40028fb..e9b34d1e1111a784e6aabd42e3724718a17d804b 100644
--- a/src/main/java/gg/pufferfish/pufferfish/compat/ServerConfigurations.java
+++ b/src/main/java/gg/pufferfish/pufferfish/compat/ServerConfigurations.java
@@ -27,7 +27,8 @@ public class ServerConfigurations {
         "spigot.yml",
         "config/paper-global.yml",
         "config/paper-world-defaults.yml",
-        "pufferfish.yml"
+        "pufferfish.yml",
+        "fish.yml"
     };
 
     private static final String[] hiddenConfigEntries = new String[]{
diff --git a/src/main/java/gg/pufferfish/pufferfish/flare/PluginLookup.java b/src/main/java/gg/pufferfish/pufferfish/flare/PluginLookup.java
index 5f4122fe66e6841c0dbca2e52b6e7ccda2317b43..c7dd02ece710b92d326ab703bfda267e2644b978 100644
--- a/src/main/java/gg/pufferfish/pufferfish/flare/PluginLookup.java
+++ b/src/main/java/gg/pufferfish/pufferfish/flare/PluginLookup.java
@@ -34,6 +34,7 @@ public class PluginLookup {
             || name.startsWith("ca.spottedleaf.") || name.startsWith("com.sun.") || name.startsWith("org.jline.") || name.startsWith("org.bukkit.")
             || name.startsWith("org.spigotmc.") || name.startsWith("com.destroystokyo.paper") || name.startsWith("co.aikar.")
             || name.startsWith("io.netty.") || name.startsWith("com.velocitypowered.")
+            || name.startsWith("me.biquaternions.fish.") // Fish - Ignore self packages
         ) {
             return Optional.empty();
         }
