From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Biquaternions <biquaternions@serlith.net>
Date: Mon, 13 Jul 2026 18:46:37 -0500
Subject: [PATCH] Fish Parallel World Ticking


diff --git a/src/main/java/gg/pufferfish/pufferfish/flare/ProfilingManager.java b/src/main/java/gg/pufferfish/pufferfish/flare/ProfilingManager.java
index da6af03d0dbfcd2c8820d5fc798f6f344307b674..35a7250f5e3ef9c41d2d004d661375d23f6c5755 100644
--- a/src/main/java/gg/pufferfish/pufferfish/flare/ProfilingManager.java
+++ b/src/main/java/gg/pufferfish/pufferfish/flare/ProfilingManager.java
@@ -101,6 +101,12 @@ public class ProfilingManager {
             GlobalMemory memory = hardware.getMemory();
             VirtualMemory virtualMemory = memory.getVirtualMemory();
 
+            // Fish start - Flare world performance collectors
+            java.util.List<co.technove.flare.live.Collector> collectors = com.google.common.collect.Lists.newArrayList(new TPSCollector(), new WorldCountCollector(ProfilingManager::submitToMainThread), new GCEventCollector(), new StatCollector(), new ThreadCollector());
+            if (me.biquaternions.fish.FishConfig.getInstance().async.worldTicking.enabled) {
+                collectors.add(me.biquaternions.fish.flare.collectors.WorldTpsCollector.create());
+            }
+            // Fish start - Flare world performance collectors
             FlareBuilder builder = new FlareBuilder()
                 .withProfileType(profileType)
                 .withMemoryProfiling(true)
@@ -112,7 +118,7 @@ public class ProfilingManager {
                 .withVersion("Minecraft Version", Bukkit.getMinecraftVersion())
 
                 .withGraphCategories(CustomCategories.ENTITIES_AND_CHUNKS, CustomCategories.MC_PERF)
-                .withCollectors(new TPSCollector(), new WorldCountCollector(ProfilingManager::submitToMainThread), new GCEventCollector(), new StatCollector(), new ThreadCollector())
+                .withCollectors(collectors.toArray(co.technove.flare.live.Collector[]::new)) // Fish - Flare world performance collectors
                 .withClassIdentifier(PluginLookup::getPluginForClass)
 
                 .withHardware(new FlareBuilder.HardwareBuilder()
