From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Biquaternions <biquaternions@serlith.net>
Date: Mon, 13 Jul 2026 18:46:37 -0500
Subject: [PATCH] Fish Parallel World Ticking


diff --git a/src/main/java/gg/pufferfish/pufferfish/flare/ProfilingManager.java b/src/main/java/gg/pufferfish/pufferfish/flare/ProfilingManager.java
index 5b907462f8998fbdefe27d7270eb5cdde548f3d2..e972eb0e30b79534fa5c8b9ba67fdd4224754658 100644
--- a/src/main/java/gg/pufferfish/pufferfish/flare/ProfilingManager.java
+++ b/src/main/java/gg/pufferfish/pufferfish/flare/ProfilingManager.java
@@ -104,6 +104,12 @@ public class ProfilingManager {
             GlobalMemory memory = hardware.getMemory();
             VirtualMemory virtualMemory = memory.getVirtualMemory();
 
+            // Fish start - Flare world performance collectors
+            java.util.List<co.technove.flare.live.Collector> collectors = com.google.common.collect.Lists.newArrayList(new TPSCollector(), new WorldCountCollector(ProfilingManager::submitToMainThread), new GCEventCollector(), new StatCollector(), new ThreadCollector());
+            if (MinecraftServer.getServer().fish$worldExecutor.isConcurrentExecutor()) {
+                collectors.add(me.biquaternions.fish.flare.collectors.WorldTpsCollector.create());
+            }
+            // Fish start - Flare world performance collectors
             FlareBuilder builder = new FlareBuilder()
                 .withProfileType(profileType)
                 .withMemoryProfiling(true)
@@ -115,7 +121,7 @@ public class ProfilingManager {
                 .withVersion("Minecraft Version", Bukkit.getMinecraftVersion())
 
                 .withGraphCategories(CustomCategories.ENTITIES_AND_CHUNKS, CustomCategories.MC_PERF)
-                .withCollectors(new TPSCollector(), new WorldCountCollector(ProfilingManager::submitToMainThread), new GCEventCollector(), new StatCollector(), new ThreadCollector())
+                .withCollectors(collectors.toArray(co.technove.flare.live.Collector[]::new)) // Fish - Flare world performance collectors
                 .withClassIdentifier(PluginLookup::getPluginForClass)
 
                 .withHardware(new FlareBuilder.HardwareBuilder()
