From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Biquaternions <biquaternions@serlith.net>
Date: Wed, 18 Jun 2025 18:23:06 -0500
Subject: [PATCH] Fish Rebrand


diff --git a/net/minecraft/CrashReport.java b/net/minecraft/CrashReport.java
index 355a5e04afe3a1f685bc5fab7e6161e01a7c69c5..9cb5b8219dd8348ca3a5156e4a46d520dfd4b6c0 100644
--- a/net/minecraft/CrashReport.java
+++ b/net/minecraft/CrashReport.java
@@ -30,7 +30,7 @@ public class CrashReport {
     private boolean trackingStackTrace = true;
     private StackTraceElement[] uncategorizedStackTrace = new StackTraceElement[0];
     private final SystemReport systemReport = new SystemReport();
-    private final List<String> extraInfo = List.of("", "DO NOT REPORT THIS TO PAPER! REPORT THIS TO PUFFERFISH!", ""); // Pufferfish - Rebrand
+    private final List<String> extraInfo = List.of("", "DO NOT REPORT THIS TO PAPER! REPORT THIS TO FISH!", ""); // Pufferfish - Rebrand // Fish - Rebrand
 
     public CrashReport(String title, Throwable exception) {
         io.papermc.paper.util.StacktraceDeobfuscator.INSTANCE.deobfuscateThrowable(exception); // Paper
diff --git a/net/minecraft/server/Eula.java b/net/minecraft/server/Eula.java
index 1f0413214f54e9a47f53c623484d6882c85f54c9..f806ec570f25d7c36eaad5f94dbee90ed9233019 100644
--- a/net/minecraft/server/Eula.java
+++ b/net/minecraft/server/Eula.java
@@ -46,7 +46,7 @@ public class Eula {
             try (OutputStream outputStream = Files.newOutputStream(this.file)) {
                 Properties map = new Properties();
                 map.setProperty("eula", "false");
-                map.store(outputStream, "By changing the setting below to TRUE you are indicating your agreement to our EULA (" + CommonLinks.EULA + ").");
+                map.store(outputStream, "By changing the setting below to TRUE you are indicating your agreement to our EULA (" + CommonLinks.EULA + ").\nYou also agree to the fish supremacy.\nReject monke, return to FISH!");
             } catch (Exception var6) {
                 LOGGER.warn("Failed to save {}", this.file, var6);
             }
diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index 094d2d528cb74b8f1d277cd780bba7f438dbe764..ac744cee72a62ef25e112a9dc6ff08489f0b373f 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -150,6 +150,17 @@ public class Main {
                 return;
             }
 
+            // Fish start - Rebrand, we just love ascii arts
+            LOGGER.info("");
+            LOGGER.info("         /`·.¸");
+            LOGGER.info("        /¸...¸`:·");
+            LOGGER.info("    ¸.·´  ¸   `·.¸.·´)   ╔═╗┬┌─┐┬ ┬");
+            LOGGER.info("   : © ):´;      ¸  {    ╠╣ │└─┐├─┤");
+            LOGGER.info("    `·.¸ `·  ¸.·´\\`·¸)   ╚  ┴└─┘┴ ┴");
+            LOGGER.info("        `\\\\´´\\¸.·´");
+            LOGGER.info("");
+            // Fish end - Rebrand, we just love ascii arts
+
             // Paper start - Detect headless JRE
             String awtException = io.papermc.paper.util.ServerEnvironment.awtDependencyCheck();
             if (awtException != null) {
diff --git a/net/minecraft/server/gui/MinecraftServerGui.java b/net/minecraft/server/gui/MinecraftServerGui.java
index f262a7c5ae4e7d56f16f5c0f4f145a2e428abbe4..f102244871fd6e96ad113afe35439467db4222ab 100644
--- a/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/net/minecraft/server/gui/MinecraftServerGui.java
@@ -79,11 +79,16 @@ public class MinecraftServerGui extends JComponent {
         this.server = server;
         this.setPreferredSize(new Dimension(854, 480));
         this.setLayout(new BorderLayout());
+        // Fish start - 🐟
+        JPanel leftPanel = new JPanel(new BorderLayout());
 
         try {
+            leftPanel.add(this.buildFishPanel(), "Center");
+            leftPanel.add(this.buildInfoPanel(), "South");
             this.add(this.buildOnboardingPanel(), "North"); // Paper - Add onboarding message for initial server start
             this.add(this.buildChatPanel(), "Center");
-            this.add(this.buildInfoPanel(), "West");
+            this.add(leftPanel, "West");
+            // Fish end - 🐟
         } catch (Exception var3) {
             LOGGER.error("Couldn't build server GUI", (Throwable)var3);
         }
@@ -110,6 +115,27 @@ public class MinecraftServerGui extends JComponent {
         return jScrollPane;
     }
 
+    // Fish start - 🐟
+    private JComponent buildFishPanel() {
+        JPanel jPanel = new JPanel(new BorderLayout());
+        try {
+            javax.swing.ImageIcon fishIcon = new javax.swing.ImageIcon(
+                java.util.Objects.requireNonNull(
+                    MinecraftServerGui.class.getClassLoader().getResource("fish.gif")
+                )
+            );
+            javax.swing.JLabel fishLabel = new javax.swing.JLabel(fishIcon);
+            jPanel.add(fishLabel, BorderLayout.CENTER);
+        } catch (net.serlith.fish.exception.FishException betrayal) {
+            LOGGER.error("Fish has been betrayed.", betrayal);
+            System.exit(1);
+        }
+
+        jPanel.setBorder(new TitledBorder(new EtchedBorder(), "Fish"));
+        return jPanel;
+    }
+    // Fish end - 🐟
+
     private JComponent buildChatPanel() {
         JPanel jPanel = new JPanel(new BorderLayout());
         JTextArea jTextArea = new JTextArea();
diff --git a/net/minecraft/world/level/chunk/storage/RegionFileStorage.java b/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
index 2b0b3b25ac28972e4852f9fbc465f5dba4a68358..af20d272ef82d79ac36bf1bea7e421ca641262c0 100644
--- a/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
+++ b/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
@@ -309,7 +309,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
 
     // Paper start
     private static void printOversizedLog(String msg, Path file, int x, int z) {
-        org.apache.logging.log4j.LogManager.getLogger().fatal(msg + " (" + file.toString().replaceAll(".+[\\\\/]", "") + " - " + x + "," + z + ") Go clean it up to remove this message. /minecraft:tp " + (x<<4)+" 128 "+(z<<4) + " - DO NOT REPORT THIS TO PUFFERFISH - You may ask for help on Discord, but do not file an issue. These error messages can not be removed."); // Pufferfish - Rebrand
+        org.apache.logging.log4j.LogManager.getLogger().fatal(msg + " (" + file.toString().replaceAll(".+[\\\\/]", "") + " - " + x + "," + z + ") Go clean it up to remove this message. /minecraft:tp " + (x<<4)+" 128 "+(z<<4) + " - DO NOT REPORT THIS TO FISH - You may ask for help on Discord, but do not file an issue. These error messages can not be removed."); // Pufferfish - Rebrand // Fish - Rebrand
     }
 
     private static CompoundTag readOversizedChunk(RegionFile regionfile, ChunkPos chunkCoordinate) throws IOException {
