From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Biquaternions <biquaternions@serlith.net>
Date: Sat, 19 Apr 2025 21:18:45 -0500
Subject: [PATCH] Fish Added Config And Command


diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index 2566e01a1fe9eb298a3b1652d93b84b118d498bb..a0536182d43d74c9225a737a49bd78eea238ee8e 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -110,6 +110,7 @@ public class Main {
                 JvmProfiler.INSTANCE.start(Environment.SERVER);
             }
 
+            me.biquaternions.fish.FishConfig.init(); // Fish - Config and command
             io.papermc.paper.plugin.PluginInitializerManager.load(options); // Paper
             Bootstrap.bootStrap();
             Bootstrap.validate();
diff --git a/net/minecraft/server/ReloadableServerResources.java b/net/minecraft/server/ReloadableServerResources.java
index 350ce5c17891198933e5c5dbcf4fa43abe8b3dc4..c070bcc808e99419c5da4f38288ce37d5edbae4c 100644
--- a/net/minecraft/server/ReloadableServerResources.java
+++ b/net/minecraft/server/ReloadableServerResources.java
@@ -47,6 +47,7 @@ public class ReloadableServerResources {
         io.papermc.paper.command.PaperCommands.registerCommands(); // Paper
         gg.pufferfish.pufferfish.PufferfishCommand.init(); // Pufferfish - Config and command
         gg.pufferfish.pufferfish.flare.FlareCommand.init(); // Pufferfish - Config and command
+        me.biquaternions.fish.FishCommand.init(); // Fish - Config and command
         this.advancements = new ServerAdvancementManager(loadingContext.lookupWithUpdatedTags());
         this.functionLibrary = new ServerFunctionLibrary(functionCompilationPermissions, this.commands.getDispatcher());
     }
