From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Biquaternions <biquaternions@serlith.net>
Date: Sat, 19 Apr 2025 21:18:45 -0500
Subject: [PATCH] Fish Added Config And Command


diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index 4a063946a3c72c16085a4db00ddd46d156c83a16..af8d7028c75fd3bcb1fa3a363b9912493578409d 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -107,6 +107,7 @@ public class Main {
                 JvmProfiler.INSTANCE.start(Environment.SERVER);
             }
 
+            new net.serlith.fish.FishConfig().load(); // Fish - Config and command
             io.papermc.paper.plugin.PluginInitializerManager.load(optionSet); // Paper
             Bootstrap.bootStrap();
             Bootstrap.validate();
diff --git a/net/minecraft/server/ReloadableServerResources.java b/net/minecraft/server/ReloadableServerResources.java
index 0265fee09afd7cded53fb606a1333974537831af..fe73f46e56e7d0545f19040d76444e391095b096 100644
--- a/net/minecraft/server/ReloadableServerResources.java
+++ b/net/minecraft/server/ReloadableServerResources.java
@@ -43,6 +43,7 @@ public class ReloadableServerResources {
         io.papermc.paper.command.PaperCommands.registerCommands(); // Paper
         gg.pufferfish.pufferfish.PufferfishCommand.init(); // Pufferfish - Config and command
         gg.pufferfish.pufferfish.flare.FlareCommand.init(); // Pufferfish - Config and command
+        net.serlith.fish.FishCommand.init(); // Fish - Config and command
         this.advancements = new ServerAdvancementManager(registries);
         this.functionLibrary = new ServerFunctionLibrary(functionCompilationLevel, this.commands.getDispatcher());
     }
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index a7c9119bb2d92b9524c1d45927091c59709179af..526d4a6ce03b02c2a409dcd4429ff91236563b7b 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -278,6 +278,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         com.destroystokyo.paper.Metrics.PaperMetrics.startMetrics(); // Paper - start metrics
         com.destroystokyo.paper.VersionHistoryManager.INSTANCE.getClass(); // Paper - load version history now
         new gg.pufferfish.pufferfish.PufferfishConfig().load(); // Pufferfish - Config and command
+        net.serlith.fish.FishConfig.init(); // Fish start - Config and command
 
         // this.worldData.setGameType(properties.gameMode.get()); // CraftBukkit - moved to world loading
         LOGGER.info("Default game type: {}", properties.gameMode.get());
