From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrPowerGamerBR <git@mrpowergamerbr.com>
Date: Sun, 25 May 2025 21:39:32 -0300
Subject: [PATCH] SparklyPaper Parallel World Ticking MSPT Tracking


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index c9d17de4e1281cb64e2a68ce118da0f00553762f..5940d51f0bde3af92ada4fa55d91de9fcca0ce42 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -2889,6 +2889,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             times = new long[length];
         }
 
+        public void _fish_add(int index, long time) { this.add(index, time); } // Fish - Expose add
         void add(int index, long time) {
             times[index % times.length] = time;
         }
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 93fcf1f43087ac66ce6ee6d715fdb6d704f528ff..9d5c20282f88ed30db16cd6757e1bd3877957612 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -566,6 +566,10 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     }
     // Paper end - chunk tick iteration
 
+    public final MinecraftServer.TickTimes tickTimes5s = new MinecraftServer.TickTimes(100);
+    public final MinecraftServer.TickTimes tickTimes10s = new MinecraftServer.TickTimes(200);
+    public final MinecraftServer.TickTimes tickTimes60s = new MinecraftServer.TickTimes(1200);
+
     public ServerLevel(
         MinecraftServer server,
         Executor dispatcher,
