From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sat, 2 Aug 2025 20:35:47 -0500
Subject: [PATCH] Leaf Fix Enderpearl Teleport


diff --git a/net/minecraft/world/entity/projectile/ThrownEnderpearl.java b/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
index 5aa63c90e1c0605faef76588efdc7c3903be4605..38df4aa1e76f2a57d2865d3190d719d60cabf5ee 100644
--- a/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
@@ -119,6 +119,33 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                 Vec3 vec3 = this.oldPosition();
                 if (owner instanceof ServerPlayer serverPlayer) {
                     if (serverPlayer.connection.isAcceptingMessages()) {
+                        // Leaf start - Fix teleport cross-dimension
+                        if (net.serlith.fish.FishConfig.ASYNC.WORLD_TICKING._ENABLED) {
+                            serverPlayer.getBukkitEntity().taskScheduler.schedule((ServerPlayer taskPlayer) -> {
+                                ServerPlayer serverPlayer1 = taskPlayer.teleport(new TeleportTransition(serverLevel, vec3, Vec3.ZERO, 0.0F, 0.0F, Relative.union(Relative.ROTATION, Relative.DELTA), TeleportTransition.DO_NOTHING, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.ENDER_PEARL));
+                                if (serverPlayer1 == null) {
+                                    this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT);
+                                    return;
+                                }
+                                if (this.random.nextFloat() < 0.05F && serverLevel.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) {
+                                    Endermite endermite = EntityType.ENDERMITE.create(serverLevel, EntitySpawnReason.TRIGGERED);
+                                    if (endermite != null) {
+                                        endermite.snapTo(owner.getX(), owner.getY(), owner.getZ(), owner.getYRot(), owner.getXRot());
+                                        serverLevel.addFreshEntity(endermite, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.ENDER_PEARL);
+                                    }
+                                }
+                                if (this.isOnPortalCooldown()) {
+                                    owner.setPortalCooldown();
+                                }
+                                if (serverPlayer1 != null) {
+                                    serverPlayer1.resetFallDistance();
+                                    serverPlayer1.resetCurrentImpulseContext();
+                                    serverPlayer1.hurtServer(serverPlayer.level(), this.damageSources().enderPearl().eventEntityDamager(this), 5.0F); // CraftBukkit // Paper - fix DamageSource API
+                                }
+                                this.playSound(serverLevel, vec3);
+                            }, e -> {}, 0);
+                        } else {
+                        // Leaf end - Fix teleport cross-dimension
                         // CraftBukkit start
                         // Store pre teleportation position as the teleport has been moved up.
                         final double preTeleportX = serverPlayer.getX(), preTeleportY = serverPlayer.getY(), preTeleportZ = serverPlayer.getZ();
@@ -155,6 +182,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                         }
 
                         this.playSound(serverLevel, vec3);
+                        } // Leaf
                     }
                 } else {
                     Entity entity = owner.teleport(
