From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sun, 3 Aug 2025 22:59:22 -0500
Subject: [PATCH] Leaf Fix Too Many Chained Neighbors


diff --git a/io/papermc/paper/redstone/RedstoneWireTurbo.java b/io/papermc/paper/redstone/RedstoneWireTurbo.java
index ff747a1ecdf3c888bca0d69de4f85dcd810b6139..342d144ce4890210216e3e16784b870fba7c24c3 100644
--- a/io/papermc/paper/redstone/RedstoneWireTurbo.java
+++ b/io/papermc/paper/redstone/RedstoneWireTurbo.java
@@ -829,14 +829,16 @@ public final class RedstoneWireTurbo {
         j = getMaxCurrentStrength(upd, j);
         int l = 0;
 
+        final int k; if (me.biquaternions.fish.FishConfig.ASYNC.WORLD_TICKING._ENABLED) k = wire.getBlockSignal(worldIn, upd.self); else {
         wire.shouldSignal = false;
         // Unfortunately, World.isBlockIndirectlyGettingPowered is complicated,
         // and I'm not ready to try to replicate even more functionality from
         // elsewhere in Minecraft into this accelerator.  So sadly, we must
         // suffer the performance hit of this very expensive call.  If there
         // is consistency to what this call returns, we may be able to cache it.
-        final int k = worldIn.getBestNeighborSignal(upd.self);
+        k = worldIn.getBestNeighborSignal(upd.self);
         wire.shouldSignal = true;
+        } // Leaf - Fix too many chained neighbors
 
         // The variable 'k' holds the maximum redstone power value of any adjacent blocks.
         // If 'k' has the highest level of all neighbors, then the power level of this
diff --git a/net/minecraft/world/level/block/RedStoneWireBlock.java b/net/minecraft/world/level/block/RedStoneWireBlock.java
index bd230046a9e9563b42d5d3c55ab49e83a1d65bde..b573ec3097950c6e76cedbc7a17405d8786c2ec7 100644
--- a/net/minecraft/world/level/block/RedStoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedStoneWireBlock.java
@@ -65,6 +65,7 @@ public class RedStoneWireBlock extends Block {
     private final BlockState crossState;
     private final RedstoneWireEvaluator evaluator = new DefaultRedstoneWireEvaluator(this);
     public boolean shouldSignal = true;
+    private final ThreadLocal<Boolean> fish$shouldSignalTL = ThreadLocal.withInitial(() -> true);
 
     @Override
     public MapCodec<RedStoneWireBlock> codec() {
@@ -329,6 +330,12 @@ public class RedStoneWireBlock extends Block {
     }
 
     public int getBlockSignal(Level level, BlockPos pos) {
+        if (me.biquaternions.fish.FishConfig.ASYNC.WORLD_TICKING._ENABLED) {
+            this.fish$shouldSignalTL.set(false);
+            int bestNeighborSignal = level.getBestNeighborSignal(pos);
+            this.fish$shouldSignalTL.set(true);
+            return bestNeighborSignal;
+        }
         this.shouldSignal = false;
         int bestNeighborSignal = level.getBestNeighborSignal(pos);
         this.shouldSignal = true;
@@ -423,18 +430,20 @@ public class RedStoneWireBlock extends Block {
 
     @Override
     protected int getDirectSignal(BlockState state, BlockGetter level, BlockPos pos, Direction side) {
+        if (me.biquaternions.fish.FishConfig.ASYNC.WORLD_TICKING._ENABLED) return !this.fish$shouldSignalTL.get() ? 0 : state.getSignal(level, pos, side);
         return !this.shouldSignal ? 0 : state.getSignal(level, pos, side);
     }
 
     @Override
     protected int getSignal(BlockState state, BlockGetter level, BlockPos pos, Direction side) {
-        if (this.shouldSignal && side != Direction.DOWN) {
+        boolean _shouldSignal = me.biquaternions.fish.FishConfig.ASYNC.WORLD_TICKING._ENABLED ? this.fish$shouldSignalTL.get() : this.shouldSignal;
+        if (_shouldSignal && side != Direction.DOWN) {
             int powerValue = state.getValue(POWER);
             if (powerValue == 0) {
                 return 0;
             } else {
                 return side != Direction.UP
-                        && !this.getConnectionState(level, state, pos).getValue(PROPERTY_BY_DIRECTION.get(side.getOpposite())).isConnected()
+                    && !this.getConnectionState(level, state, pos).getValue(PROPERTY_BY_DIRECTION.get(side.getOpposite())).isConnected()
                     ? 0
                     : powerValue;
             }
@@ -460,6 +469,7 @@ public class RedStoneWireBlock extends Block {
 
     @Override
     protected boolean isSignalSource(BlockState state) {
+        if (me.biquaternions.fish.FishConfig.ASYNC.WORLD_TICKING._ENABLED) return this.fish$shouldSignalTL.get();
         return this.shouldSignal;
     }
 
