From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Biquaternions <biquaternions@serlith.net>
Date: Sat, 18 Jul 2026 15:47:07 -0500
Subject: [PATCH] Fish Handle Entity Scheduler


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 55a579e48a9bb9db77da0d648b079b429afcc8ac..daf411e2e3df30746a64080d051414462eb14fbc 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1772,6 +1772,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         this.server.getScheduler().mainThreadHeartbeat(); // CraftBukkit
         // Paper start - optimise Folia entity scheduler
         ((io.papermc.paper.threadedregions.scheduler.FoliaGlobalRegionScheduler)org.bukkit.Bukkit.getGlobalRegionScheduler()).tick();
+        if (!me.biquaternions.fish.FishConfig.ASYNC.WORLD_TICKING._ENABLED) { // Fish - Handle Entity Schedulers
         for (io.papermc.paper.threadedregions.EntityScheduler scheduler : this.entitySchedulerTickList.getAllSchedulers()) {
             if (scheduler.isRetired()) {
                 continue;
@@ -1779,6 +1780,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
             scheduler.executeTick();
         }
+        } // Fish - Handle Entity Schedulers
         // Paper end - optimise Folia entity scheduler
         // Paper end - Folia scheduler API
         io.papermc.paper.adventure.providers.ClickCallbackProviderImpl.ADVENTURE_CLICK_MANAGER.handleQueue(this.tickCount); // Paper
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 7fdabf824574bb3e48779c0b1a07e700ffa3f2dd..7567b6c13d908dd9fbaba302cfc863730f6bceb3 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -5259,6 +5259,10 @@ public abstract class Entity
     // Paper end - Folia schedulers
     // Paper start - optimise Folia entity scheduler
     public final void registerScheduler() {
+        if (me.biquaternions.fish.FishConfig.ASYNC.WORLD_TICKING._ENABLED && this.level instanceof ServerLevel serverLevel) {
+            this.getBukkitEntity().taskScheduler.registerTo(serverLevel.fish$worldData.entitySchedulerTickList);
+            return;
+        }
         this.getBukkitEntity().taskScheduler.registerTo(net.minecraft.server.MinecraftServer.getServer().entitySchedulerTickList);
     }
     // Paper end - optimise Folia entity scheduler
